File: app/src/main/java/com/tosh/poolandroid/LoginRegistration/LoginActivity.java
package com.tosh.poolandroid.LoginRegistration;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.tosh.poolandroid.MainActivity;
import com.tosh.poolandroid.R;
import com.tosh.poolandroid.Retrofit.AuthRetrofitClient;
import com.tosh.poolandroid.Retrofit.NodeAuth;

import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;

import static android.util.Patterns.EMAIL_ADDRESS;

public class LoginActivity extends AppCompatActivity {

    MaterialButton registerBtn;
    MaterialButton loginBtn;
    MaterialButton forgotBtn;

    TextInputEditText inputPassword, inputEmail;

    NodeAuth api;
    CompositeDisposable compositeDisposable = new CompositeDisposable();


    @Override
    protected void onStop() {
        compositeDisposable.clear();
        super.onStop();
    }

    @Override
    protected void onDestroy() {
        compositeDisposable.clear();
        super.onDestroy();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        //init api
        Retrofit retrofit = AuthRetrofitClient.getInstance();
        api = retrofit.create(NodeAuth.class);

        // views
        inputEmail =  findViewById(R.id.email_login);
        inputPassword = findViewById(R.id.password_login);
        loginBtn = findViewById(R.id.login_btn);
        forgotBtn = findViewById(R.id.btn_forgotPassword);
        registerBtn = findViewById(R.id.login_register_btn);

        registerBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LoginActivity.this, RegisterActivity.class);
                startActivity(intent);
                finish();
            }
        });

        loginBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loginUser(inputEmail.getText().toString(), inputPassword.getText().toString());
            }
        });
    }

    private void loginUser(String email, String password) {

        if(TextUtils.isEmpty(email)){
            inputEmail.setError("Enter email");
            return;
        }
        if(!EMAIL_ADDRESS.matcher(email).matches()){
            inputEmail.setError("Enter valid email");
            return;
        }
        if(TextUtils.isEmpty(password)){
            inputPassword.setError("Enter password");
            return;
        }
        compositeDisposable.add(api.loginUser(email, password)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Consumer<String>() {
            @Override
            public void accept(String res) throws Exception {

                System.out.println("YESSSSSSSSSSSSSSSSSSSSSSSSSSSS"+ res);

                if (res .equals("successful")){
                    Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                    startActivity(intent);
                    finish();
                }else {
                    Toast.makeText(LoginActivity.this, ""+res, Toast.LENGTH_SHORT).show();
                }
            }
        })
        );

    }

}

File: app/src/main/java/com/tosh/poolandroid/view/activity/LoginActivity.kt
package com.tosh.poolandroid.view.activity

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.WindowManager
import android.view.animation.DecelerateInterpolator
import android.view.animation.Interpolator
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.agrawalsuneet.dotsloader.loaders.LazyLoader
import com.google.android.material.snackbar.Snackbar
import com.tosh.poolandroid.R
import com.tosh.poolandroid.viewmodel.MainViewModel
import kotlinx.android.synthetic.main.activity_login.*



class LoginActivity : AppCompatActivity() {

    lateinit var containerLL: LinearLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        containerLL = findViewById(R.id.containerLL) as LinearLayout

        this.window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)

        login_register_btn.setOnClickListener {
            val intent = Intent(applicationContext, RegisterActivity::class.java)

            startActivity(intent)
            finish()
        }

        login_btn.setOnClickListener {
            val email = email_login.text.toString().trim()
            val password = password_login.text.toString().trim()

            if (email.isEmpty()){
                email_login.error = "Enter email"
                return@setOnClickListener
            }

            if (password.isEmpty()){
                password_login.error = "Enter password"
                return@setOnClickListener
            }

            loading()
            instantiateLoginViewModel(email, password)

        }
    }

    private fun instantiateLoginViewModel(email:String,  password:String) {
        containerLL.visibility = VISIBLE
        val mainViewModel: MainViewModel = ViewModelProviders.of(this)[MainViewModel::class.java]
        mainViewModel.userLogin(email, password).observe(this, Observer {

            if (it == "successful"){
                val intent = Intent(applicationContext, MainActivity::class.java)
                addToSharedPreferences(email)
                startActivity(intent)
                finish()
            }else{
                val contextView: View = findViewById(R.id.loginLayout)
                val snackbar = Snackbar
                    .make(contextView, it, Snackbar.LENGTH_LONG)
                snackbar.show()
                containerLL.visibility = GONE
            }
        })
    }

    fun addToSharedPreferences(email: String) {
        var preferences: SharedPreferences? = PreferenceManager.getDefaultSharedPreferences(this)
        var editor: SharedPreferences.Editor?

        editor = preferences?.edit()
        editor?.putString("email", email)
        editor?.apply()

    }

    private fun loading(){
        var lazyLoader = LazyLoader(this, 15, 5,
            ContextCompat.getColor(this, R.color.loader_selected),
            ContextCompat.getColor(this, R.color.loader_selected),
            ContextCompat.getColor(this, R.color.loader_selected))
            .apply {
                animDuration = 500
                firstDelayDuration = 100
                secondDelayDuration = 200
                interpolator = DecelerateInterpolator() as Interpolator
            }
        containerLL.addView(lazyLoader)
    }
}
